classes/api/opensesame.php

```diff
@@ -0,0 +1,286 @@
+     */
+    private function get_request($resource, $header = [], $params = [], $url = '') {
+        if (empty($url)) {
+            $url = $this->baseurl . $resource;
+        }
+
+        !PHPUNIT_TEST ?? mtrace("Getting this: $url");
+        $this->resetHeader();
+        $this->resetcookie();
+        $this->reset_request_state_vars();
+        $this->resetopt();
+        $header = array_merge($header, $this->get_authentication_header());
+        $this->setHeader($header);
+
+        $response = $this->get($url, $params);
+
+        $httpcode = $this->get_http_code();
+        if ($httpcode !== 200) {
+            if ($httpcode === 401) {
+                unset_config('accesstoken', 'tool_opensesame');
+                $this->accesstoken = null;
+            }
+            throw new \Exception("Could not process GET HTTP request for API URL $url: " . $this->get_http_code());
+        }
+
+        return $response;
+    }
+
+    /**
+     * Executes a GET request with a specific set of retry attempts.
+     * If 200 code is not achieved, it will retry until exhausting the allowed amount.
+     * @param string $resource
+     * @param array $header
+     * @param array $params
+     * @param string $url
+     * @return string
+     */
+    private function get_request_retries($resource, $header = [], $params = [], $url = '') {
+        $maxattempts = $this->retries;
+        $attempts = 0;
+        $response = false;
+        $exception = null;
+        do {
+            try {
+                $response = $this->get_request($resource, $header, $params, $url);
+                if ($response !== false) {
+                    break;
+                } else {
+                    sleep(3);
+                }
+            } catch (\Exception $ex) {
+                $exception = $ex;
+                sleep(3);
+            } finally {
+                $attempts += 1;
+            }
+        } while ($attempts < $maxattempts && $response === false);
+
+        if ($resource === false && !is_null($exception)) {
+            throw $exception;
+        }
+        return $response;
+    }
+
+    /**
+     * Authenticates with the Opensesame auth URL to get a token.
+     * @return bool true if auth succeeded.
+     * @throws \dml_exception
+     */
+    private function authenticate(): bool {
+        !PHPUNIT_TEST ?? mtrace("Authenticating with Open Sesame");
+        $clientid = $this->clientid;
+        $clientsecret = $this->clientsecret;
+        $this->resetHeader();
+        $this->setHeader([
+            'Content-Type: application/x-www-form-urlencoded',
+            'Accept: application/json',
+            sprintf('Authorization: Basic %s', base64_encode(sprintf('%s:%s', $clientid, $clientsecret))),
+        ]);
+        $authurl = $this->authurl;
+        $response = $this->post($authurl, 'grant_type=client_credentials&scope=content');
+        if ($this->get_http_code() !== 200) {
+            !PHPUNIT_TEST ?? mtrace($response);
```

Comment:
Does this mean that if we are NOT executing unit tests, the full response will be logged?

Reply:
It is helful when the response is not valid, but we can delete it if you want to.

Reply:
And yep, logs on unitests are not useful on a pipeline, it is better not to show them.

Reply:
I definitely want to keep the log messages!  Thanks for the clarification